src/librustc_resolve/late/diagnostics.rs

```diff
@@ -507,15 +513,74 @@ impl<'a> LateResolutionVisitor<'a, '_, '_> {
                             ),
                         );
                     }
-                    suggested = true;
                 }
-                _ => {}
-            }
-            if !suggested {
-                if let Some(span) = self.r.opt_span(def_id) {
-                    err.span_label(span, &format!("`{}` defined here", path_str));
+                PathSource::TupleStruct(span) => {
+                    err.set_span(span);
+                    if let Some(span) = self.r.opt_span(def_id) {
+                        err.span_label(span, &format!("`{}` defined here", path_str));
+                    }
+                    let (fields, appl) = match self.r.field_names.get(&def_id) {
+                        Some(fields) => (
+                            fields
+                                .iter()
+                                .map(|f| f.node.to_string())
+                                .collect::<Vec<String>>()
+                                .join(", "),
+                            Applicability::MachineApplicable,
+                        ),
+                        None => ("/* fields */".to_string(), Applicability::HasPlaceholders),
+                    };
+                    err.span_suggestion(
+                        span,
+                        "use struct pattern syntax instead",
+                        format!("{} {{{pad}{}{pad}}}", path_str, fields, pad = pad),
+                        appl,
+                    );
+                }
+                PathSource::Expr(Some(Expr {
+                    span: call_span,
+                    kind: ExprKind::Call(expr, _),
+                    ..
+                })) if expr.span == span => {
+                    err.set_span(*call_span);
+                    if let Some(span) = self.r.opt_span(def_id) {
+                        err.span_label(span, &format!("`{}` defined here", path_str));
+                    }
+                    let fields = match self.r.field_names.get(&def_id) {
+                        Some(fields) => fields
+                            .iter()
+                            .map(|f| format!("{}: val", f.node))
+                            .collect::<Vec<String>>()
+                            .join(", "),
+                        None => "/* fields */".to_string(),
+                    };
+                    err.span_suggestion(
+                        *call_span,
+                        "use struct literal syntax instead",
+                        format!("{} {{{pad}{}{pad}}}", path_str, fields, pad = pad),
+                        Applicability::HasPlaceholders,
+                    );
+                }
+                PathSource::Pat if matches!(self.diagnostic_metadata.current_expr, Some(_)) => {
```

Comment:
`current_expr` is always `None` in this PR and can be removed.

Even if it were set like it is set in https://github.com/rust-lang/rust/pull/74210, the condition would still be meaningless - "path pattern is (indirectly) inside of some expression" applies to a large subset of path patterns just because blocks are expressions.

Reply:
`current_expr` in `https://github.com/rust-lang/rust/pull/74210` doesn't do the expected thing either.
You probably need to embed a flag into `PathSource::Type` that is true only if the `TyKind::Path` in question is directly embedded into `ExprKind::Type`.

Reply:
Changed for this PR, I'll look at the other one in a bit.